File: api/typescript/src/engine/lib/kurtosis_context/enclave_identifiers.ts
import {EnclaveUUID} from "../../../core/lib/enclaves/enclave_context";
import kurtosis_engine_rpc_api_bindings =  require("../../kurtosis_engine_rpc_api_bindings/engine_service_pb");

const VALID_UUID_MATCHES_ALLOWED = 1

// Docs available at https://docs.kurtosis.com/sdk#enclave-identifiers
export class EnclaveIdentifiers {
    public readonly enclaveNameToUuids: Map<string, EnclaveUUID[]>;
    public readonly enclaveUuids: Map<EnclaveUUID, boolean>;
    public readonly enclaveShortenedUuidToUuids: Map<string, EnclaveUUID[]>;

    constructor(historicalIdentifiers : kurtosis_engine_rpc_api_bindings.EnclaveIdentifiers[]) {
        this.enclaveUuids = new Map<string, boolean>();
        this.enclaveNameToUuids = new Map<string, EnclaveUUID[]>();
        this.enclaveShortenedUuidToUuids = new Map<string, EnclaveUUID[]>();
        historicalIdentifiers.forEach((enclaveIdentifiers) => {
            let enclaveName = enclaveIdentifiers.getName();
            let enclaveUuid : EnclaveUUID = enclaveIdentifiers.getEnclaveUuid();
            let shortenedUuid = enclaveIdentifiers.getShortenedUuid();

            this.enclaveUuids.set(enclaveUuid, true);

            if (!(enclaveName in this.enclaveNameToUuids)) {
                this.enclaveNameToUuids.set(enclaveName, []);
            }
            let enclaveUuids = this.enclaveNameToUuids.get(enclaveName)!
            enclaveUuids.push(enclaveUuid)
            this.enclaveNameToUuids.set(enclaveName, enclaveUuids)

            if (!(shortenedUuid in this.enclaveShortenedUuidToUuids)) {
                this.enclaveShortenedUuidToUuids.set(shortenedUuid, []);
            }
            enclaveUuids = this.enclaveShortenedUuidToUuids.get(shortenedUuid)!
            enclaveUuids.push(enclaveUuid)
            this.enclaveShortenedUuidToUuids.set(shortenedUuid, enclaveUuids)
        });
    }

    public getEnclaveUuidForIdentifier(identifier: string): EnclaveUUID {
        if (this.enclaveUuids.has(identifier)) {
            return identifier as EnclaveUUID
        }

        if (this.enclaveShortenedUuidToUuids.has(identifier)) {
            let matches = this.enclaveShortenedUuidToUuids.get(identifier)!
            if (matches.length === VALID_UUID_MATCHES_ALLOWED) {
                return matches[0];
            } else if (matches.length > VALID_UUID_MATCHES_ALLOWED) {
                throw new Error(`Found multiple enclaves ${matches} matching shortened uuid ${identifier}. Please use a uuid to be more specific`)
            }
        }

        if (this.enclaveNameToUuids.has(identifier)) {
            let matches = this.enclaveNameToUuids.get(identifier)!
            if (matches.length === VALID_UUID_MATCHES_ALLOWED) {
                return matches[0];
            } else if (matches.length > VALID_UUID_MATCHES_ALLOWED) {
                throw new Error(`Found multiple enclaves ${matches} matching name ${identifier}. Please use a uuid to be more specific`)
            }
        }

        throw new Error(`No matching uuid for identifier ${identifier}`)
    }

    public getOrderedListOfNamesAndUuids(): String[] {
        let enclaveNames: string[] = [];
        let enclaveUuids: string[] = [];

        for (let name in this.enclaveNameToUuids) {
            enclaveNames.push(name);
        }

        for (let uuid in this.enclaveUuids) {
            enclaveUuids.push(uuid);
        }

        enclaveNames.sort();
        enclaveUuids.sort();

        return [...enclaveNames, ...enclaveUuids];
    }
}

File: api/golang/engine/lib/kurtosis_context/enclave_identifiers.go
package kurtosis_context

import (
	"github.com/kurtosis-tech/kurtosis/api/golang/v2/v2/core/lib/enclaves"
	"github.com/kurtosis-tech/kurtosis/api/golang/v2/v2/engine/kurtosis_engine_rpc_api_bindings"
	"github.com/kurtosis-tech/stacktrace"
	"sort"
)

// Docs available at https://docs.kurtosis.com/sdk#enclave-identifiers
type EnclaveIdentifiers struct {
	enclaveNameToUuids          map[string][]enclaves.EnclaveUUID
	enclaveUuids                map[enclaves.EnclaveUUID]bool
	enclaveShortenedUuidToUuids map[string][]enclaves.EnclaveUUID
}

func newEnclaveIdentifiers(allIdentifiers []*kurtosis_engine_rpc_api_bindings.EnclaveIdentifiers) *EnclaveIdentifiers {
	enclaveNames := map[string][]enclaves.EnclaveUUID{}
	enclaveUuids := map[enclaves.EnclaveUUID]bool{}
	enclaveShortenedUuid := map[string][]enclaves.EnclaveUUID{}
	for _, enclaveIdentifiers := range allIdentifiers {
		enclaveName := enclaveIdentifiers.GetName()
		enclaveUuid := enclaves.EnclaveUUID(enclaveIdentifiers.EnclaveUuid)
		enclaveUuids[enclaveUuid] = true
		shortenedUuid := enclaveIdentifiers.GetShortenedUuid()
		enclaveNames[enclaveName] = append(enclaveNames[enclaveName], enclaveUuid)
		enclaveShortenedUuid[shortenedUuid] = append(enclaveShortenedUuid[shortenedUuid], enclaveUuid)
	}

	return &EnclaveIdentifiers{enclaveNameToUuids: enclaveNames, enclaveUuids: enclaveUuids, enclaveShortenedUuidToUuids: enclaveShortenedUuid}
}

func (identifiers *EnclaveIdentifiers) GetEnclaveUuidForIdentifier(identifier string) (enclaves.EnclaveUUID, error) {
	if _, found := identifiers.enclaveUuids[enclaves.EnclaveUUID(identifier)]; found {
		return enclaves.EnclaveUUID(identifier), nil
	}

	if matches, found := identifiers.enclaveShortenedUuidToUuids[identifier]; found {
		if len(matches) == validUuidMatchesAllowed {
			return matches[0], nil
		} else if len(matches) > validUuidMatchesAllowed {
			return "", stacktrace.NewError("Found multiple enclaves '%v' matching shortened uuid '%v'. Please use a uuid to be more specific", matches, identifier)
		}
	}

	if matches, found := identifiers.enclaveNameToUuids[identifier]; found {
		if len(matches) == validUuidMatchesAllowed {
			return matches[0], nil
		} else if len(matches) > validUuidMatchesAllowed {
			return "", stacktrace.NewError("Found multiple enclaves '%v' matching name '%v'. Please use a uuid to be more specific", matches, identifier)
		}
	}

	return "", stacktrace.NewError("No matching uuid for identifier '%s'", identifier)
}

func (identifiers *EnclaveIdentifiers) GetOrderedListOfNames() []string {
	var enclaveNames []string

	for name := range identifiers.enclaveNameToUuids {
		enclaveNames = append(enclaveNames, name)
	}

	sort.Strings(enclaveNames)
	return enclaveNames
}
